### PR TITLE
Don't create async boundaries for SyncIO#bracket

### LINKED ISSUE #529: IO#unsafeRunSync erroneously fails on JavaScript
Consider this program:

```scala
def safe: Unit = IO.unit.guarantee(IO.unit).unsafeRunSync
def broken = IO.unit.bracket(_ => IO(safe))(_ => IO.unit) 
```

Calling `safe` on Scala.js correctly results in a `Unit` value. However, calling `safe` from within  another nested `IO` program (e.g. `broken` here), results in the following error:

```
[info]   java.lang.UnsupportedOperationException: cannot synchronously await result on JavaScript; use runAsync or unsafeRunAsync
[info]   at java.lang.UnsupportedOperationException.fillInStackTrace(/Users/mpilquist/Development/oss/fs2/core/js/target/scala-2.12/fs2-core-test-fastopt.js:76968:14)
[info]   at java.lang.UnsupportedOperationException.<init>(/Users/mpilquist/Development/oss/fs2/core/js/target/scala-2.12/fs2-core-test-fastopt.js:77090:10)
[info]   at java.lang.UnsupportedOperationException.<init>(/Users/mpilquist/Development/oss/fs2/core/js/target/scala-2.12/fs2-core-test-fastopt.js:132545:58)
[info]   at cats.effect.internals.IOPlatform$.unsafeResync(/Users/mpilquist/Development/oss/fs2/core/js/target/scala-2.12/fs2-core-test-fastopt.js:7111:53)
[info]   at cats.effect.IO$Async.unsafeRunTimed(/Users/mpilquist/Development/oss/fs2/core/js/target/scala-2.12/fs2-core-test-fastopt.js:47886:52)
[info]   at cats.effect.IO$Async.unsafeRunSync(/Users/mpilquist/Development/oss/fs2/core/js/target/scala-2.12/fs2-core-test-fastopt.js:47843:15)
[info]   at fs2.StreamSpec.safe$1(/Users/mpilquist/Development/oss/fs2/core/js/target/scala-2.12/fs2-core-test-fastopt.js:227516:106)
[info]   at {anonymous}()(/Users/mpilquist/Development/oss/fs2/core/js/target/scala-2.12/fs2-core-test-fastopt.js:208627:23)
[info]   at scala.scalajs.runtime.AnonFunction0.apply(/Users/mpilquist/Development/oss/fs2/core/js/target/scala-2.12/fs2-core-test-fastopt.js:97704:23)
[info]   at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(/Users/mpilquist/Development/oss/fs2/core/js/target/scala-2.12/fs2-core-test-fastopt.js:7376:25)
```

I believe the cause is related to the internal trampoline used when evaluating the outer program.

### PR DESCRIPTION
**Pull request by LukaJCB**
Should fix #529 

Before: 
```scala
scala> def safe: Unit = SyncIO.unit.guarantee(SyncIO.unit).unsafeRunSync
safe: Unit

scala> def broken = SyncIO.unit.bracket(_ => SyncIO(safe))(_ => SyncIO.unit)
broken: cats.effect.SyncIO[Unit]

scala> broken.unsafeRunSync
java.lang.UnsupportedOperationException: cannot synchronously await result on JavaScript; use runAsync or unsafeRunAsync
...
```

After:
```scala
scala> def safe: Unit = SyncIO.unit.guarantee(SyncIO.unit).unsafeRunSync
safe: Unit

scala> def broken = SyncIO.unit.bracket(_ => SyncIO(safe))(_ => SyncIO.unit)
broken: cats.effect.SyncIO[Unit]

scala> broken.unsafeRunSync

scala>
```

### CODE REVIEW

File: core/shared/src/main/scala/cats/effect/SyncIO.scala [hunk 1]
```diff
@@ -175,7 +175,11 @@ final class SyncIO[+A] private (val toIO: IO[A]) {
    *        along with the result of `use` (error or successful result)
    */
   def bracketCase[B](use: A => SyncIO[B])(release: (A, ExitCase[Throwable]) => SyncIO[Unit]): SyncIO[B] =
-    new SyncIO(toIO.bracketCase(a => use(a).toIO)((a, ec) => release(a, ec).toIO))
+    flatMap { a => use(a).attempt.flatMap {
```
Comment (alexandru) [inline core/shared/src/main/scala/cats/effect/SyncIO.scala]: You might want to use `redeemWith` instead of `attempt.flatMap`.

Comment (LukaJCB) [reply core/shared/src/main/scala/cats/effect/SyncIO.scala]: Good call 👍

**Comment by alexandru** [comment]
I think we might need to change that `toIO` conversion as well 😕

**Comment by alexandru** [comment]
I'm afraid that we're now forced to do this:

```scala
def toIO[A](fa: SyncIO[A]): IO[A] = IO.delay(fa.unsafeRunSync)
```

Which we'd probably do anyway, with a custom implementation of `SyncIO`.

**Comment by LukaJCB** [comment]
Hmm, will that be binary compatible?

**Comment by alexandru** [comment]
Or we could make it `uncancelable` on conversion.

What do you think?

**Comment by alexandru** [comment]
I think we can change it — the constructor is private and we could define our own `toIO`, as a `def` maybe?

IDK, see what Mima tells you.

**Comment by LukaJCB** [comment]
Mima seems fine with it :)

**Comment by mpilquist** [comment]
Looks good to me, good find!

**Pull request merged**